### PR TITLE
Keep order for set operators such as UNION / UNION ALL

### PR DESCRIPTION
**Pull request by csthomas**
Pull Request for Issue #115

### Summary of Changes
1. Fix UNION/UNION ALL
2. Allow to mix UNION with UNION ALL and preserve the order.
3. Allow to support (in the future) more set operators like `EXCEPT` and `INTERSECT`
4. A few B/C breaks
    - Method `unionDistinct()` - removed, use `union()` instead.
    - Protected variables `$union` and `$unionAll` replaced by variable `$merge` (array) to preserve the order.
    - Array is not accepted in the $query parameter of `union()` and `unionAll()` methods.
4. New methods `DatabaseQuery::querySet($query)` and `DatabaseQuery::toQuerySet()`, see https://github.com/joomla-framework/database/pull/116#issuecomment-385015308
### Testing Instructions

1. Go to `database/src` folder.
2. Run `php -a` or create a short script like below and execute it. Replace dots `...`.

```php
require '../vendor/autoload.php';
$db = Joomla\Database\DatabaseDriver::getInstance(['driver' => 'mysqli', 'host' => 'localhost', 'user' => '...', 'password' => '...', 'database' => '...', 'prefix' => '...']);
print_r($db->setQuery($db->getQuery(true)->select('1')->union($db->getQuery(true)->select('2'))->unionAll($db->getQuery(true)->select('3')))->loadColumn());
Array
(
    [0] => 1
    [1] => 2
    [2] => 3
)
print_r($db->setQuery($db->getQuery(true)->select('2')->union($db->getQuery(true)->select('1'))->unionAll($db->getQuery(true)->select('1')))->loadColumn());
Array
(
    [0] => 2
    [1] => 1
    [2] => 1
)
print_r($db->setQuery($db->getQuery(true)->select('2')->unionAll($db->getQuery(true)->select('1'))->union($db->getQuery(true)->select('1')))->loadColumn());
Array
(
    [0] => 2
    [1] => 1
)
```

##### Changes in methods `union()`, `unionAll()` and `unionDistinct()`

- Method `unionDistinct()` has been removed. Use `union()` instead.
- Argument `$query` stops accepting the array. Only `DatabaseQuery` object or string.
- Argument `$distinct` has been removed from `unionAll`.
- Argument `$glue` has been removed from both.

Method `union()` by default has `$distinct = true`.
If `$distinct` is `false` then generates `UNION ALL` sql statement.

Class variables `$union` and `$unionAll` has been merged into one variable `$merge`.
The new variable represents an ordered array of individual elements.

Stop supporting `$type = 'union'` in method `__toString()`.

### Mix of union types
```sql
SELECT 1 UNION SELECT 2 UNION ALL SELECT 1 -- > return 3 rows

SELECT 1 UNION ALL SELECT 2 UNION SELECT 1 -- > return 2 rows
```

> You can combine multiple queries using the set operators UNION, UNION ALL, INTERSECT, and MINUS. All set operators have equal precedence. If a SQL statement contains multiple set operators, then Oracle Database evaluates them from the left to right unless parentheses explicitly specify another order.

Source https://docs.oracle.com/cd/B19306_01/server.102/b14200/queries004.htm

### CODE REVIEW

**Comment by mbabker** [comment]
Seems fair to me.  Just add a note to https://github.com/joomla-framework/database/blob/2.0-dev/docs/v1-to-v2-update.md and I'd say this should be fine, we just need to sort out deprecation notices then in the CMS repo and this repo's master (1.x) branch.

**Comment by csthomas** [comment]
If necessary, correct my documentation.

**Comment by alikon** [comment]
look nice to me too, IIRC `UNION` in core is used only in com_finder

**Comment by csthomas** [comment]
Nice, I think about it "too much" and now I have another solution, where we can put ORDER BY and LIMIT before UNION statement. I will prepare another PR for that.

**Comment by alikon** [comment]
:+1: 
despite i'm not a fan of "LIMIT" clause on such kind of analytics query :fearful:

File: docs/v1-to-v2-update.md [hunk 1]
```diff
@@ -58,3 +58,20 @@ Database drivers now support dispatching read-only events when a connection to t
...
+
+- Method `unionDistinct()` has been removed. Use `union()` instead.
+- Argument `$query` stops accepting the array. Only `DatabaseQuery` object or string.
+- Argument `$distinct` has been removed from `unionAll`.
+- Argument `$glue` has been removed from both.
+
+Method `union()` by default has `$distinct = true`.
+If `$distinct` is `false` then generates `UNION ALL` sql statement.
+
+Class variables `$union` and `$unionAll` has been merged into one variable `$merge`.
```
Comment (Quy) [inline docs/v1-to-v2-update.md]: Change `has` to `have`

**Comment by csthomas** [comment]
At the end I added new unit tests for PostgreSQL and SQLite. 

SqlSrv may still does not work if you use ORDER BY or LIMIT in union query. This is more complicated fix and I would not add it to this PR.

What do you thing about this? @mbabker  @alikon @Quy

**Comment by mbabker** [comment]
Seems fine to me.

> SqlSrv may still does not work if you use ORDER BY or LIMIT in union query. This is more complicated fix and I would not add it to this PR.

Well luckily the testing infrastructure here is a lot better off (and you can somewhat copy the current Travis config to get a working SQL Server instance to test against).  No tests are failing for now so let's roll with it and we can work on updating the test suite to better cover cases like this.

**Comment by simbunch** [comment]
Tested successfully.

**Comment by csthomas** [comment]
I will spend a little time to fix sqlsrv stuff, I need to back to my previous PR #104

**Comment by csthomas** [comment]
My tests:

```php
require 'vendor/autoload.php';
$db = Joomla\Database\DatabaseDriver::getInstance(['driver' => 'sqlsrv', 'database' => '', 'user' => '', 'password' => '', 'prefix' => 'j38_']);
$q = $db->getQuery(1)->select('id')->from('j38_modules')->where('id <= 5')->order('id DESC')->setLimit(2, 1)->toQuerySet()->unionAll($db->getQuery(1)->select('id')->from('j38_modules')->where('id = 3 OR id = 4')->order('0+id DESC')->setLimit(1, 1))->order('id DESC');

echo $q;

SELECT * FROM (
SELECT * FROM (
SELECT * FROM (
SELECT *, ROW_NUMBER() OVER (ORDER BY (SELECT 0)) AS RowNumber_4c1bcee9724ef8f6f37135d9596d299a
FROM (
SELECT TOP 3 id
FROM j38_modules
WHERE id <= 5
ORDER BY id DESC) AS A
) AS A
WHERE RowNumber_4c1bcee9724ef8f6f37135d9596d299a > 1) AS merge_0
UNION ALL SELECT * FROM (
SELECT * FROM (
SELECT *, ROW_NUMBER() OVER (ORDER BY (SELECT 0)) AS RowNumber_b0af95c67b1589f1db7914ea2e26f658
FROM (
SELECT TOP 2 id
FROM j38_modules
WHERE id = 3 OR id = 4
ORDER BY 0+id DESC) AS A
) AS A
WHERE RowNumber_b0af95c67b1589f1db7914ea2e26f658 > 1) AS merge_1
) AS merges
ORDER BY id DESC

echo $q->setLimit(5, 2);

SELECT * FROM (
SELECT *, ROW_NUMBER() OVER (ORDER BY (SELECT 0)) AS RowNumber_e89d72665d4b94533c8b495e1eebf7ee
FROM (
SELECT TOP 7 * FROM (
SELECT * FROM (
SELECT * FROM (
SELECT *, ROW_NUMBER() OVER (ORDER BY (SELECT 0)) AS RowNumber_4c1bcee9724ef8f6f37135d9596d299a
FROM (
SELECT TOP 3 id
FROM j38_modules
WHERE id <= 5
ORDER BY id DESC) AS A
) AS A
WHERE RowNumber_4c1bcee9724ef8f6f37135d9596d299a > 1) AS merge_0
UNION ALL SELECT * FROM (
SELECT * FROM (
SELECT *, ROW_NUMBER() OVER (ORDER BY (SELECT 0)) AS RowNumber_b0af95c67b1589f1db7914ea2e26f658
FROM (
SELECT TOP 2 id
FROM j38_modules
WHERE id = 3 OR id = 4
ORDER BY 0+id DESC) AS A
) AS A
WHERE RowNumber_b0af95c67b1589f1db7914ea2e26f658 > 1) AS merge_1
) AS merges
ORDER BY id DESC) AS A
) AS A
WHERE RowNumber_e89d72665d4b94533c8b495e1eebf7ee > 2
```

It looks a little ugly but we can only remove `ROW_NUMBER()` after we bump to SQL Server 2012

**Comment by csthomas** [comment]
Now unit tests execute real query tests for UNIONs on `sqlite`, `mysqli`, `pgsql` and `sqlsrv`.

**Comment by csthomas** [comment]
Unit tests passed.

File: src/Pdo/PdoDriver.php [hunk 2]
```diff
@@ -681,7 +681,7 @@ protected function prepareStatement(string $query): StatementInterface
 		}
 		catch (\PDOException $exception)
 		{
-			throw new PrepareStatementFailureException($exception->getMessage(), $exception->getCode(), $exception);
+			throw new PrepareStatementFailureException($exception->getMessage() . $query, (int) $exception->getCode(), $exception);
```
[the remark this answers is not in the document]
Comment (csthomas) [reply src/Pdo/PdoDriver.php]: Ups, it was only for debug, I will remove it

**Comment by csthomas** [comment]
> Also, we shouldn't cast the exception code to integer here. It is correct that the DB drivers don't strictly use numeric codes and Exception codes aren't type enforced.

Yes, but try to do a bug in code:
```diff
diff --git a/src/Sqlite/SqliteQuery.php b/src/Sqlite/SqliteQuery.php
index 940d030..b196fcc 100644
--- a/src/Sqlite/SqliteQuery.php
+++ b/src/Sqlite/SqliteQuery.php
@@ -73,7 +73,7 @@ class SqliteQuery extends PdoQuery
                        case 'querySet':
                                $query = $this->querySet;
 
-                               if ($query->order || $query->limit || $query->offset)
+                               if (0)//$query->order || $query->limit || $query->offset)
```
 
Then run phpunit by `php phpunit Tests/Sqlite/SqliteDriverTest.php`

Then I got:

1) Joomla\Database\Tests\Sqlite\SqliteDriverTest::testQuerySetWithUnionAll
Error: Wrong parameters for Joomla\Database\Exception\PrepareStatementFailureException([string $message [, long $code [, Throwable $previous = NULL]]])

Maybe override constructor for `PrepareStatementFailureException` could be better?

**Comment by mbabker** [comment]
If it really comes down to that yes the constructor should be overloaded.  Now I'm curious what is overloading core PHP's `RuntimeException`, or even root `Exception`, definition because last I was unaware the root constructor was not typed.

**Comment by csthomas** [comment]
```php
php7.1 -a
Interactive mode enabled

php > new RuntimeException('', '');
PHP Warning:  Uncaught Error: Wrong parameters for RuntimeException([string $message [, long $code [, Throwable $previous = NULL]]]) in php shell code:1
Stack trace:
#0 php shell code(1): Exception->__construct('', '')
#1 {main}
  thrown in php shell code on line 1
```

**Comment by mbabker** [comment]
:eyeroll:

**Comment by csthomas** [comment]
It seems to be a known issue http://php.net/manual/en/class.pdoexception.php#95812

Should I add the constructor for `PrepareStatementFailureException` at this PR?

**Comment by mbabker** [comment]
Might as well.

**Comment by csthomas** [comment]
Done

**Pull request merged**